File: pkg/wfcdgen/wfcdgen.go
// Package wfcdgen は WFCD の生アイテムデータ（pkg/wfcd）から、Chain View に追加できる
// ノード候補を組み立てる。項目10の設計方針どおり「自動確定」ではなく「候補をサジェストして
// 人が選ぶ」レベルに意図的にスコープを絞っている — 特にパーツ→レリックはOR関係（どれか1つで
// 良い）の候補が複数あるため、ユーザーが1つ選んでから requires に追加する前提で、
// このパッケージは選択肢の提示までしか行わない（AND前提の既存engineパッケージはそのままでよい）。
package wfcdgen

import (
	"regexp"
	"strings"

	"warframe-state-graph/pkg/model"
	"warframe-state-graph/pkg/wfcd"
)

type Paradigm string

const (
	// 03_Data_Source_Research.md 2.8/2.10/2.11節でまとめた装備系入手構造の6パラダイム。
	ParadigmSingleBlueprint Paradigm = "single-blueprint" // ①本体設計図のみ型
	ParadigmMultiPart       Paradigm = "multi-part"       // ②パーツ複数型（Prime武器/フレーム/アークウィング等）
	ParadigmModular         Paradigm = "modular"          // ③モジュール自由組立型（Zaw/Kitgun/Amp/Moa/Hound/K-Drive）
	ParadigmInstant         Paradigm = "instant"          // ④即時完成品型（componentsが存在しない）
	ParadigmFrameAssociated Paradigm = "frame-associated" // ⑤フレーム付随型（Exalted Weapon/Venari）
	ParadigmBreeding        Paradigm = "breeding"         // ⑥DNA/繁殖型（Kubrow/Kavat等）
)

// modularTypes は Misc.json の type フィールドがこれらの場合、③モジュール型と分類する。
var modularTypes = map[string]bool{
	"Zaw Component":    true,
	"Kitgun Component": true,
	"Amp":              true,
}

// ClassifyParadigm はアイテム1件をどのパラダイムに分類するかの推定。components/type/category
// だけから判定する簡易ヒューリスティックであり（正確な分類にはFoundryレシピの有無等の
// 追加情報が必要）、最終的な精度はユーザーが候補を見て取捨選択する前提で許容している。
func ClassifyParadigm(item wfcd.Item) Paradigm {
	if modularTypes[item.Type] {
		return ParadigmModular
	}
	if len(item.Components) == 0 {
		if item.Category == wfcd.CategoryPets {
			return ParadigmBreeding
		}
		return ParadigmInstant
	}
	// 複数の別々のドロップ元を持つコンポーネントが2つ以上あれば「各パーツが個別入手先を持つ」
	// パーツ複数型と判定する。単一Blueprint+汎用素材（駒素材はdropsを持たないことが多い）は
	// ①single-blueprintのまま。
	partsWithOwnDrop := 0
	for _, c := range item.Components {
		if len(c.Drops) > 0 {
			partsWithOwnDrop++
		}
	}
	if partsWithOwnDrop >= 2 {
		return ParadigmMultiPart
	}
	return ParadigmSingleBlueprint
}

// richLichPrefixes は2.6節で実データ検証済みの識別方法（名前プレフィックス一致が唯一100%網羅）。
var richLichPrefixes = []string{"Kuva ", "Tenet ", "Coda "}

// DetectRichLich は Kuva/Tenet/Coda リッチ系武器かどうかを名前プレフィックスで判定する。
func DetectRichLich(name string) (kind string, ok bool) {
	for _, p := range richLichPrefixes {
		if strings.HasPrefix(name, p) {
			return strings.TrimSpace(p), true
		}
	}
	return "", false
}

// Archetype は武器の基礎ステータスから見た大まかな設計思想。
type Archetype string

const (
	ArchetypeCrit    Archetype = "Crit"
	ArchetypeStatus  Archetype = "Status"
	ArchetypeHybrid  Archetype = "Hybrid"
	ArchetypeUtility Archetype = "Utility"
)

// 閾値はWarframeコミュニティで一般的に使われる目安（クリティカル率25%以上でクリット武器と
// 見なす等）であり、DE公式の厳密な定義ではないヒューリスティック。02_Requirements_and_Roadmap.md
// item2で「どの組み合わせが理論上最強か」ではなく「アーキタイプ判定まで」とスコープを絞った
// 方針に沿い、目安判定として十分という前提。
const (
	critChanceThreshold = 0.25
	procChanceThreshold = 0.25
)

// DetectArchetype は criticalChance/procChance の基礎値からアーキタイプを判定する。
func DetectArchetype(item wfcd.Item) Archetype {
	crit := item.CriticalChance >= critChanceThreshold
	status := item.ProcChance >= procChanceThreshold
	switch {
	case crit && status:
		return ArchetypeHybrid
	case crit:
		return ArchetypeCrit
	case status:
		return ArchetypeStatus
	default:
		return ArchetypeUtility
	}
}

// rivenStatArchetype は Riven のポジ値ステータス名がどのアーキタイプに属するかの対応表。
var rivenStatArchetype = map[string]Archetype{
	"Critical Chance": ArchetypeCrit,
	"Critical Damage": ArchetypeCrit,
	"Status Chance":   ArchetypeStatus,
	"Status Duration": ArchetypeStatus,
	"Multishot":       ArchetypeHybrid, // 両アーキタイプに寄与するためどちらとも一致扱い
	"Damage":          ArchetypeHybrid,
}

// RivenStatChoices はRiven専用入力UIでポジ値として選択できるステータス名一覧。
// アーキタイプ判定に関与するもの（rivenStatArchetypeのキー）に加え、実際のRivenで
// よく出る代表的なステータスも選択肢として含める（後者は一致判定には影響しない中立枠）。
var RivenStatChoices = []string{
	"Critical Chance", "Critical Damage", "Status Chance", "Status Duration",
	"Multishot", "Damage",
	"Fire Rate", "Reload Speed", "Punch Through", "Range", "Magazine Capacity", "Recoil",
}

// RivenCheck は武器アーキタイプとRivenのポジ値ステータス群を照合した結果。
type RivenCheck struct {
	Archetype    Archetype `json:"archetype"`
	Matches      bool      `json:"matches"`
	MatchedStats []string  `json:"matchedStats,omitempty"`
}

// CheckRiven は「そのRivenのポジ値が武器のアーキタイプと噛み合っているか」を判定する。
// 02_Requirements_and_Roadmap.md item2で確定したスコープ通り、理論値レンジの精密計算は行わず
// あくまで一致マーク表示のためのショーケース判定。
func CheckRiven(item wfcd.Item, positiveStats []string) RivenCheck {
	archetype := DetectArchetype(item)
	var matched []string
	for _, stat := range positiveStats {
		if a, ok := rivenStatArchetype[stat]; ok && (a == archetype || a == ArchetypeHybrid) {
			matched = append(matched, stat)
		}
	}
	return RivenCheck{Archetype: archetype, Matches: len(matched) > 0, MatchedStats: matched}
}

// slugPattern はノードID生成時に許可する文字（英数字とハイフン）以外を除去するためのもの。
var slugPattern = regexp.MustCompile(`[^a-z0-9]+`)

// Slug は "Ash Prime Neuroptics" のような名前を "ash-prime-neuroptics" のようなID形式に変換する。
// 既存 data/graph.json の命名規則（ash-prime, dragon-nikana-riven 等）に合わせている。
func Slug(name string) string {
	s := strings.ToLower(name)
	s = slugPattern.ReplaceAllString(s, "-")
	return strings.Trim(s, "-")
}

// RelicCandidate はパーツ1つに対する入手先レリックの候補。OR関係（どれか1つで良い）なので
// requiresへの追加は自動化せず、ユーザーが選んだ1件だけをフロント側で追加する想定。
type RelicCandidate struct {
	Name    string  `json:"name"`
	Chance  float64 `json:"chance"`
	Vaulted bool    `json:"vaulted"`
}

// PartSuggestion はパーツ1件分のノード候補＋その入手先レリック候補一覧。
type PartSuggestion struct {
	Node            *model.Node      `json:"node"`
	RelicCandidates []RelicCandidate `json:"relicCandidates,omitempty"`
}

// Suggestion は1アイテムに対するノード自動生成の提案一式。
type Suggestion struct {
	Paradigm  Paradigm  `json:"paradigm"`
	RichLich  string    `json:"richLich,omitempty"`
	Archetype Archetype `json:"archetype,omitempty"`

	Root  *model.Node      `json:"root"`            // Build/Weapon/Frame本体ノード
	Parts []PartSuggestion `json:"parts,omitempty"` // ②パーツ複数型の場合のみ
}

// relicNamePattern は Drop.Location からレリック名部分だけを拾う（例:
// "Void Relic (Axi A22) (25.33%)" のような表記から "Axi A22" を抜き出す）。
// locationの正確なフォーマットはWFCDのバージョンで変わりうるため、緩めのパターンで
// 部分一致させ、見つからない場合はlocation全体をそのまま候補名として扱う。
var relicInLocationPattern = regexp.MustCompile(`(Lith|Meso|Neo|Axi) [A-Z]\d{1,2}`)

func extractRelicName(location string) string {
	if m := relicInLocationPattern.FindString(location); m != "" {
		return m
	}
	return location
}

// BuildSuggestion は1アイテムからノード生成候補を組み立てる。activeRelicsが空/nilの場合は
// Vault判定をfalse固定にする（レリックデータ取得に失敗しても他の提案は成立させるため）。
func BuildSuggestion(item wfcd.Item, nodeType model.NodeType, activeRelics map[string]bool) *Suggestion {
	paradigm := ClassifyParadigm(item)
	richLich, _ := DetectRichLich(item.Name)

	root := &model.Node{
		ID:         Slug(item.Name),
		Name:       item.Name,
		Type:       nodeType,
		UniqueName: item.UniqueName,
	}

	sug := &Suggestion{Paradigm: paradigm, RichLich: richLich, Root: root}
	if nodeType == model.TypeWeapon {
		sug.Archetype = DetectArchetype(item)
	}

	if paradigm != ParadigmMultiPart {
		return sug
	}

	for _, c := range item.Components {
		partNode := &model.Node{
			ID:   root.ID + "-" + Slug(c.Name),
			Name: c.Name,
			Type: model.TypeResource,
		}
		root.Contains = append(root.Contains, partNode.ID)

		var candidates []RelicCandidate
		for _, d := range c.Drops {
			relicName := extractRelicName(d.Location)
			candidates = append(candidates, RelicCandidate{
				Name:    relicName,
				Chance:  d.Chance,
				Vaulted: activeRelics != nil && wfcd.IsRelicVaulted(activeRelics, relicName),
			})
		}
		sug.Parts = append(sug.Parts, PartSuggestion{Node: partNode, RelicCandidates: candidates})
	}
	return sug
}

File: ts/server/wfcdgen.ts
// Port of pkg/wfcdgen. Builds Chain View node candidates from raw WFCD item
// data (pkg/wfcd). Deliberately scoped to "suggest candidates, human picks"
// rather than "auto-confirm" (item 10) — part->relic in particular is an OR
// relationship (any one candidate is enough) with several options, so this
// package only presents choices; the caller adds exactly one chosen relic to
// `requires` (the existing AND-premised engine package needs no change).

import type { Node, NodeType } from "./model.ts";
import { MainQuestNames, Prerequisites, ResolveChain, Slug as questchainSlug } from "./questchain.ts";
import type { Item, SyndicateEntry } from "./wfcd.ts";
import { findSyndicateWeaponRank, isRelicVaulted, CategoryPets, RELIC_ERA_PREFIX } from "./wfcd.ts";

export type Paradigm =
  | "single-blueprint" // (1) blueprint-only
  | "multi-part" // (2) multi-part (Prime weapons/frames/archwings etc.)
  | "modular" // (3) freely-assembled modular (Zaw/Kitgun/Amp/Moa/Hound/K-Drive)
  | "instant" // (4) instant-complete (no components)
  | "frame-associated" // (5) frame-tied (Exalted Weapon/Venari)
  | "breeding" // (6) DNA/breeding (Kubrow/Kavat etc.)
  // quest-chain is a 7th paradigm used only for quests — a different axis
  // (prerequisite chaining) from the equipment-acquisition-structure
  // classification, but reuses the Suggestion type below.
  | "quest-chain";

// Misc.json `type` values classified as (3) modular.
const MODULAR_TYPES = new Set(["Zaw Component", "Kitgun Component", "Amp"]);

/** A simple heuristic from components/type/category alone (accurate
 * classification would need extra info like Foundry recipe presence) — the
 * human reviewing candidates before import is expected to compensate for
 * any misclassification. */
export function classifyParadigm(item: Item): Paradigm {
  if (item.type && MODULAR_TYPES.has(item.type)) return "modular";
  if (!item.components || item.components.length === 0) {
    if (item.category === CategoryPets) return "breeding";
    return "instant";
  }
  // 2+ components each with their own distinct drop source implies
  // "each part is separately acquired" (multi-part). A single blueprint +
  // generic materials (crafting materials usually carry no drops of their
  // own) stays (1) single-blueprint.
  const partsWithOwnDrop = item.components.filter((c) => c.drops && c.drops.length > 0).length;
  if (partsWithOwnDrop >= 2) return "multi-part";
  return "single-blueprint";
}

// Verified as the only ~100%-reliable identification method (name-prefix match).
const RICH_LICH_PREFIXES = ["Kuva ", "Tenet ", "Coda "];

export function detectRichLich(name: string): { kind: string; ok: boolean } {
  for (const p of RICH_LICH_PREFIXES) {
    if (name.startsWith(p)) return { kind: p.trim(), ok: true };
  }
  return { kind: "", ok: false };
}

export type Archetype = "Crit" | "Status" | "Hybrid" | "Utility";

// Thresholds are the common community rule-of-thumb (e.g. 25%+ crit chance
// counts as a "crit weapon"), not an official DE definition — item 2 scoped
// this to "archetype classification," not "which combo is theoretically
// optimal," so a rule-of-thumb suffices.
const CRIT_CHANCE_THRESHOLD = 0.25;
const PROC_CHANCE_THRESHOLD = 0.25;

export function detectArchetype(item: Item): Archetype {
  const crit = (item.criticalChance ?? 0) >= CRIT_CHANCE_THRESHOLD;
  const status = (item.procChance ?? 0) >= PROC_CHANCE_THRESHOLD;
  if (crit && status) return "Hybrid";
  if (crit) return "Crit";
  if (status) return "Status";
  return "Utility";
}

// Which archetype each Riven positive stat name belongs to.
const RIVEN_STAT_ARCHETYPE: Record<string, Archetype> = {
  "Critical Chance": "Crit",
  "Critical Damage": "Crit",
  "Status Chance": "Status",
  "Status Duration": "Status",
  Multishot: "Hybrid", // contributes to both archetypes, so it matches either
  Damage: "Hybrid",
};

/** Selectable positive-stat names for the Riven input UI. Includes the
 * archetype-matching stats above plus other commonly-rolled stats (the
 * latter are neutral — they never affect the match verdict). */
export const RivenStatChoices = [
  "Critical Chance",
  "Critical Damage",
  "Status Chance",
  "Status Duration",
  "Multishot",
  "Damage",
  "Fire Rate",
  "Reload Speed",
  "Punch Through",
  "Range",
  "Magazine Capacity",
  "Recoil",
];

export interface RivenCheck {
  archetype: Archetype;
  matches: boolean;
  matchedStats?: string[];
}

/** Whether a Riven's positive stats line up with the weapon's archetype —
 * per item 2's confirmed scope this is a showcase match indicator only, not
 * a precise theoretical-range calculation. */
export function checkRiven(item: Item, positiveStats: string[]): RivenCheck {
  const archetype = detectArchetype(item);
  const matched: string[] = [];
  for (const stat of positiveStats) {
    const a = RIVEN_STAT_ARCHETYPE[stat];
    if (a && (a === archetype || a === "Hybrid")) matched.push(stat);
  }
  return { archetype, matches: matched.length > 0, matchedStats: matched.length ? matched : undefined };
}

const SLUG_PATTERN = /[^a-z0-9]+/g;

/** "Ash Prime Neuroptics" -> "ash-prime-neuroptics", matching the existing
 * data/graph.json naming convention. */
export function Slug(name: string): string {
  const s = name.toLowerCase().replace(SLUG_PATTERN, "-");
  return s.replace(/^-+|-+$/g, "");
}

// WFCD's Component.Drops.Location mixes relic-sourced strings (e.g. "Void
// Relic (Axi B1) (25.33%)") with plain-mission-sourced ones (e.g.
// "Pluto/Fenton's Field (Skirmish), Rotation A"). When isRelic is false,
// vaulted is always false (meaningless) — a prior version didn't
// distinguish this and called isRelicVaulted on non-relic strings too,
// mislabeling plain-mission drops as "Vaulted" just because they weren't in
// the active-relics list (found 2026-08-23 from a real frame's "this text
// looks wrong" report).
export interface RelicCandidate {
  name: string;
  chance: number;
  isRelic: boolean;
  vaulted: boolean;
}

export interface PartSuggestion {
  node: Node;
  relicCandidates?: RelicCandidate[];
}

export interface SyndicateRankSuggestion {
  node: Node; // e.g. id="red-veil-exalted", name="Red Veil - Exalted"
  standing: number; // purchase cost at that rank, not cumulative standing to reach it
}

export interface Suggestion {
  paradigm: Paradigm;
  richLich?: string;
  archetype?: Archetype;

  root: Node; // the Build/Weapon/Frame node itself, or the quest's own starting node
  parts?: PartSuggestion[]; // only for (2) multi-part

  // Only set for paradigm "quest-chain": every node from the earliest
  // prerequisite through root (root included, no duplicates). Each node's
  // requires is already set to its prerequisite's node id (questchain's Slug convention).
  questChain?: Node[];

  // Only set for a syndicate weapon (Vaykor/Secura/Rakta/Synoid/Telos/Sancti
  // etc.). Not auto-added to root.requires (same as parts — the human
  // confirms at import time, per item 10's "suggest, don't auto-confirm" policy).
  syndicateRank?: SyndicateRankSuggestion;
}

// Loosely matches a relic name inside a Drop.Location (e.g. pulls "Axi A22"
// out of "Void Relic (Axi A22) (25.33%)"). WFCD's location format can shift
// between versions, so this partial-matches rather than parsing strictly. No
// match means the drop isn't relic-sourced (plain mission/assassination
// drop) — returning isRelic=false here matters (see RelicCandidate's comment above).
// Built from wfcd.ts's RELIC_ERA_PREFIX (not a second hardcoded era list) —
// see that constant's comment for why (2026-08-27, the Vanguard fix).
const RELIC_IN_LOCATION_PATTERN = new RegExp(`(${RELIC_ERA_PREFIX}) [A-Z]\\d{1,2}`);

function extractRelicName(location: string): { name: string; isRelic: boolean } {
  const m = RELIC_IN_LOCATION_PATTERN.exec(location);
  if (m) return { name: m[0], isRelic: true };
  return { name: location, isRelic: false };
}

/** Builds a node-generation suggestion for one item. An empty/undefined
 * activeRelics fixes vault status to false (so a relic-data fetch failure
 * doesn't block the rest of the suggestion). An empty/undefined syndicates
 * skips the syndicate-rank suggestion, same best-effort policy. */
export function BuildSuggestion(
  item: Item,
  nodeType: NodeType,
  activeRelics: Set<string> | undefined,
  syndicates: Record<string, SyndicateEntry[]> | undefined,
): Suggestion {
  const paradigm = classifyParadigm(item);
  const { kind: richLich, ok: hasRichLich } = detectRichLich(item.name);

  const root: Node = {
    id: Slug(item.name),
    name: item.name,
    type: nodeType,
    satisfied: false,
    requires: [],
    contains: [],
    uniqueName: item.uniqueName,
  };

  const sug: Suggestion = { paradigm, root };
  if (hasRichLich) sug.richLich = richLich;
  if (nodeType === "Weapon") {
    sug.archetype = detectArchetype(item);
    if (syndicates) {
      const rank = findSyndicateWeaponRank(syndicates, item.name);
      if (rank) {
        const rankNode: Node = {
          id: `${Slug(rank.syndicate)}-${Slug(rank.rankLabel)}`,
          name: `${rank.syndicate} - ${rank.rankLabel}`,
          type: "Syndicate",
          satisfied: false,
          requires: [],
          contains: [],
        };
        sug.syndicateRank = { node: rankNode, standing: rank.standing };
      }
    }
  }

  if (paradigm !== "multi-part") return sug;

  const parts: PartSuggestion[] = [];
  for (const c of item.components ?? []) {
    const partNode: Node = {
      id: `${root.id}-${Slug(c.name)}`,
      name: c.name,
      type: "Resource",
      satisfied: false,
      requires: [],
      contains: [],
    };
    root.contains!.push(partNode.id);

    const candidates: RelicCandidate[] = [];
    for (const d of c.drops ?? []) {
      const { name, isRelic } = extractRelicName(d.location);
      const cand: RelicCandidate = { name, chance: d.chance, isRelic, vaulted: false };
      if (isRelic) cand.vaulted = activeRelics != null && isRelicVaulted(activeRelics, name);
      candidates.push(cand);
    }
    parts.push({ node: partNode, relicCandidates: candidates.length ? candidates : undefined });
  }
  sug.parts = parts;
  return sug;
}

/** Walks questchain.MainStoryChain from questName and builds the full node
 * set (earliest prerequisite through questName, in order). A quest name not
 * registered in MainStoryChain (a side quest etc.) returns just itself
 * (no prerequisites) — out-of-table quests are treated as "prerequisite unknown". */
export function BuildQuestSuggestion(questName: string): Suggestion {
  const chain = ResolveChain(questName);
  const nodes: Node[] = chain.map((name) => ({
    id: questchainSlug(name),
    name,
    type: "Quest",
    satisfied: false,
    requires: Prerequisites(name).map((pre) => questchainSlug(pre)),
    contains: [],
  }));
  const root = nodes[nodes.length - 1]!; // ResolveChain returns prerequisites-before-subject, so the last entry is the starting quest.
  return { paradigm: "quest-chain", root, questChain: nodes };
}

// Re-exported so callers (main.ts) don't need a separate questchain import
// just for the main-quest-name list used alongside wfcdgen elsewhere.
export { MainQuestNames };
